File: app/src/test/java/com/mss/instamat/repositories/network/ImagesNetRepositoryImplTest.java
package com.mss.instamat.repositories.network;

import com.mss.instamat.domain.repositories.ImagesNetRepository;
import com.mss.instamat.repositories.network.models.ImageInfoNet;
import com.mss.instamat.repositories.network.models.ImagesResponse;

import org.jetbrains.annotations.NotNull;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Maybe;
import io.reactivex.android.plugins.RxAndroidPlugins;
import io.reactivex.schedulers.Schedulers;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class ImagesNetRepositoryImplTest {

    private ImagesNetRepository repository;

    @Mock
    private PixabayAPI pixabayAPI;

    @BeforeClass
    public static void setupClass() {
        RxAndroidPlugins.setInitMainThreadSchedulerHandler(__ -> Schedulers.trampoline());
    }

    @Before
    public void setUp() {
        repository = new ImagesNetRepositoryImpl(pixabayAPI);
    }

    @Test
    public void findImages_callAPIAndReturnsMappedImages() {
        final ImagesResponse imagesResponse = getImagesResponse();
        when(pixabayAPI.findImages(anyString(), anyString(), anyString(), anyString(), anyInt(), anyInt())).thenReturn(Maybe.just(imagesResponse));

        assertEquals(imagesResponse.getHits().size(), repository.findImages("one", 1).blockingGet().size());

        verify(pixabayAPI).findImages(anyString(), anyString(), anyString(), anyString(), anyInt(), anyInt());
    }

    @NotNull
    private ImagesResponse getImagesResponse() {
        ImagesResponse imagesResponse = new ImagesResponse();
        ImageInfoNet imageInfoNet = new ImageInfoNet();
        List<ImageInfoNet> imagesNet = new ArrayList<>();
        imagesNet.add(imageInfoNet);
        imagesResponse.setHits(imagesNet);
        return imagesResponse;
    }
}

File: app/src/test/java/com/mss/imagesearcher/model/repositories/network/ImagesNetRepositoryImplTest.kt
package com.mss.imagesearcher.model.repositories.network

import com.mss.imagesearcher.model.repositories.ImagesNetRepository
import com.mss.imagesearcher.model.repositories.network.entity.ImageInfoNet
import com.mss.imagesearcher.model.repositories.network.entity.ImagesResponse
import io.reactivex.Maybe
import io.reactivex.android.plugins.RxAndroidPlugins
import io.reactivex.schedulers.Schedulers
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class ImagesNetRepositoryImplTest {

    private var repository: ImagesNetRepository? = null

    @Mock
    private val pixabayAPI: PixabayAPI? = null

    private val imagesResponse: ImagesResponse
        get() {
            val imagesResponse = ImagesResponse()
            val imageInfoNet = ImageInfoNet()
            val imagesNet = ArrayList<ImageInfoNet>()
            imagesNet.add(imageInfoNet)
            imagesResponse.hits = imagesNet
            return imagesResponse
        }

    @Before
    fun setUp() {
        repository = ImagesNetRepositoryImpl(pixabayAPI!!)
    }

    @Test
    fun findImages_callAPIAndReturnsMappedImages() {
        val imagesResponse = imagesResponse
        `when`(pixabayAPI!!.findImages(anyString(), anyString(), anyString(), anyString(), anyInt(), anyInt())).thenReturn(Maybe.just(imagesResponse))

        assertEquals(imagesResponse.hits!!.size.toLong(), repository!!.findImages("one", 1).blockingGet().size.toLong())

        verify(pixabayAPI).findImages(anyString(), anyString(), anyString(), anyString(), anyInt(), anyInt())
    }

    companion object {

        @BeforeClass
        fun setupClass() {
            RxAndroidPlugins.setInitMainThreadSchedulerHandler { Schedulers.trampoline() }
        }
    }
}